pkg/controller/certificaterequests/selfsigned/selfsigned.go

```diff
@@ -0,0 +1,173 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package selfsigned
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificaterequests"
+	crutil "github.com/jetstack/cert-manager/pkg/controller/certificaterequests/util"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	cmerrors "github.com/jetstack/cert-manager/pkg/util/errors"
+	"github.com/jetstack/cert-manager/pkg/util/kube"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	CRControllerName = "certificaterequests-issuer-selfsigned"
+)
+
+var (
+	errorNoAnnotation = fmt.Errorf("self signed issuer requires %q annotation to be set to the name of the Secret containing the private key",
+		v1alpha1.CRPrivateKeyAnnotationKey)
+)
+
+type SelfSigned struct {
+	// used to record Events about resources to the API
+	recorder record.EventRecorder
+
+	issuerOptions controllerpkg.IssuerOptions
+	secretsLister corelisters.SecretLister
+}
+
+func init() {
+	// create certificate request controller for selfsigned issuer
+	controllerpkg.Register(CRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		selfsigned := NewSelfSigned(ctx)
+		controller := certificaterequests.New(apiutil.IssuerSelfSigned, selfsigned)
+
+		c, err := controllerpkg.New(ctx, CRControllerName, controller)
+		if err != nil {
+			return nil, err
+		}
+
+		return c.Run, nil
+	})
+}
+
+func NewSelfSigned(ctx *controllerpkg.Context) *SelfSigned {
+	return &SelfSigned{
+		recorder:      ctx.Recorder,
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+	}
+}
+
+func (s *SelfSigned) Sign(ctx context.Context, cr *v1alpha1.CertificateRequest, issuerObj v1alpha1.GenericIssuer) (*issuer.IssueResponse, error) {
+	log := logf.FromContext(ctx, "sign")
+	reporter := crutil.NewReporter(cr, s.recorder)
+
+	resourceNamespace := s.issuerOptions.ResourceNamespace(issuerObj)
+
+	secretName, ok := cr.ObjectMeta.Annotations[v1alpha1.CRPrivateKeyAnnotationKey]
+	if !ok || secretName == "" {
+		message := fmt.Sprintf("Annotation %q missing or reference empty",
+			v1alpha1.CRPrivateKeyAnnotationKey)
+		err := errors.New("secret name missing")
+
+		reporter.Pending(err, "MissingAnnotation", message)
```

Comment:
Undecided: should we hard fail if the annotation is missing? Do we want to allow changes here after the request has been created? I am leaning towards hard failing here 🙄

Reply:
Yeah that's my bad - I agree, should hard fail